overlord/devicestate/handlers_install.go

```diff
@@ -1214,14 +1262,115 @@ func (m *DeviceManager) doInstallFinish(t *state.Task, _ *tomb.Tomb) error {
+	st.Unlock()
+	sys, snapInfos, snapSeeds, err := m.loadSystemAndEssentialSnaps(systemLabel,
+		[]snap.Type{snap.TypeKernel, snap.TypeBase, snap.TypeGadget})
+	st.Lock()
+	if err != nil {
+		return err
+	}
+	// Mount gadget and kernel
+	mntPtForType := make(map[snap.Type]string)
+	for _, seedSn := range []*seed.Snap{snapSeeds[snap.TypeGadget], snapSeeds[snap.TypeKernel]} {
+		mntPt, rest, err := mountSeedSnap(seedSn)
+		mntPtForType[seedSn.EssentialType] = mntPt
+		if err != nil {
+			return err
+		}
+		defer func() {
+			errRest := rest()
+			if errRest != nil {
+				err = errRest
+			}
+		}()
+	}
+
+	// unset revision means the snap will be local, we do something
+	// similar in the seedwriter
+	// XXX: is this the right place?
+	// TODO: also do this for gadget
+	kernelInfo := snapInfos[snap.TypeKernel]
+	if kernelInfo.Revision.Unset() {
+		kernelInfo.Revision = snap.R(-1)
+	}
+
+	// TODO useEncryption equals true case
+	useEncryption := false
+	installObserver, trustedInstallObserver, err := buildInstallObserver(sys.Model, mntPtForType[snap.TypeGadget], useEncryption)
+	if err != nil {
+		return err
+	}
+
+	logger.Noticef("writing content to partitions")
+	timings.Run(perfTimings, "install-finish", "Writing content to partitions", func(tm timings.Measurer) {
+		st.Unlock()
+		defer st.Lock()
+		err = installWriteContent(onVolumes, installObserver,
+			mntPtForType[snap.TypeGadget], mntPtForType[snap.TypeKernel], sys.Model, perfTimings)
+	})
+	if err != nil {
+		return fmt.Errorf("cannot write content: %v", err)
+	}
+
+	recovBootWith := &boot.RecoverySystemBootableSet{
+		Kernel:          snapInfos[snap.TypeKernel],
+		KernelPath:      snapSeeds[snap.TypeKernel].Path,
+		GadgetSnapOrDir: snapSeeds[snap.TypeGadget].Path,
+	}
+	bootWith := &boot.BootableSet{
+		Base:       snapInfos[snap.TypeBase],
+		BasePath:   snapSeeds[snap.TypeBase].Path,
+		Kernel:     snapInfos[snap.TypeKernel],
+		KernelPath: snapSeeds[snap.TypeKernel].Path,
+		Gadget:     snapInfos[snap.TypeGadget],
+		GadgetPath: snapSeeds[snap.TypeGadget].Path,
+
+		UnpackedGadgetDir: mntPtForType[snap.TypeGadget],
+
+		RecoverySystemLabel: systemLabel,
+		// TODO: important but a bit unclear why needed
+		RecoverySystemDir: systemLabel,
+
+		Recovery: true,
+
+		// XXX: rename to something more neutral like "TargetDir" ?
+		InstallHostWritableDir: filepath.Join(boot.InitramfsRunMntDir, "ubuntu-data"),
+	}
+
+	// install the bootloader configuration from the gadget
+	// TODO where to take this from?
+	seedDir := boot.InitramfsUbuntuSeedDir
+	// installs in recovery: grub.cfg, grubenv, but in install mode?
+	logger.Debugf("making the recovery partition bootable")
+	if err := boot.MakeBootableImage(sys.Model, seedDir, bootWith, boot.ModeRun, []string{}); err != nil {
+		return err
+	}
+
+	// sets environment
+	logger.Debugf("making the recovery partition runnable")
+	if err := boot.MakeRecoverySystemBootable(seedDir, "/", recovBootWith); err != nil {
+		return err
+	}
```

Comment:
the feel like abuse, especially the second one passing "/" as we have no recovery systems and also I think it writes some files that we will then just ignore?.  It seems we need a different helper that does the part of these two that we really